### DISCUSSION TITLE
Indication of invalid overloads

### DISCUSSION THREAD

**Discussion by tkukushkin** [microsoft/pyright#3123] [Q&A] [answered]
I am trying to add type stubs for class like:

```python
from typing import Generic, TypeVar, overload, Any, Sequence, Literal
from typing_extensions import Never

T = TypeVar('T')

class Serializer(Generic[T]):

    def __init__(self, dataclass: type[T]) -> None:
        self.dataclass = dataclass

    @overload
    def dump(  # single object without explicit many (default False)
        self,
        obj: T,
    ) -> dict[str, Any]: ...

    @overload
    def dump(  # single object with many=False
        self,
        obj: T,
        *,
        many: Literal[False],
    ) -> dict[str, Any]: ...

    @overload
    def dump(  # single object with many=True, result is Never!!!!
        self,
        obj: T,
        *,
        many: Literal[True],
    ) -> Never: ...

    @overload
    def dump(  # sequence with many=True
        self,
        obj: Sequence[T],  # 
        *,
        many: Literal[True],
    ) -> list[dict[str, Any]]: ...

    @overload
    def dump(  # sequence with many=False, result is Never!!!!
        self,
        obj: Sequence[T],
        *,
        many: Literal[False],
    ) -> Never: ...

    @overload
    def dump(  # sequence without explicit many (default False), result is Never!!!!
        self,
        obj: Sequence[T],
    ) -> Never: ...

    @overload
    def dump(  # sequence or not, result is dict or list.
        self,
        obj: T | Sequence[T],
        *,
        many: bool = False,
    ) -> dict[str, Any] | list[dict[str, Any]]: ...

    def dump(
        self,
        obj: T | Sequence[T],
        *,
        many: bool = False,
    ) -> dict[str, Any] | list[dict[str, Any]]:
        return {}

class A:
    pass

serializer = Serializer(A)
result = serializer.dump(A(), many=True)  # I want to get some pyright error here, such signature is invalid
reveal_type(result)
```

Because of the last overload, which is needed for cases where method parameters are somehow formed dynamically, I am also obliged to do some overloads for cases where method arguments are exactly invalid in combination. I tried using the `Never` type. pyright outputs `unknown` in this case, that's generally fine with me, but mypy even with `--strict` thinks that the result type is `Any`, which I don't really like, because that would let the user accidentally miss the error and work with `Any` without even knowing about it.

Maybe I'm doing something wrong and there's a better way to do what I want?

----
UPD: Perhaps it would be better to use `NoReturn` in this case?

**Comment by JelleZijlstra** [comment, marked as answer]
I don't think there's a good way to do this at the moment. You'd need something like python/typing#1043.

Never and NoReturn mean the same thing. They aren't a good fit here because they make the type checker treat the rest of the function as unreachable, instead of throwing an error.

Mypy's behavior with overload resolution is different from pyright's; it is more likely to fall back to Any if multiple overloads match.

**Reply by erictraut** [reply]
Yeah, I don't see a good way to do this. As @JelleZijlstra said, `Never` does not have the correct semantics, so I wouldn't recommend using it in this manner.

There has been some discussion of supporting a standardized `@deprecated` and `@error` decorator which might be useful in this case to tag specific overloads as "error conditions".

**Reply by tkukushkin** [reply]
What could be the problem with using `NoReturn` in this case? Well, except that the error will not explicitly tell me where the problem is.

**Reply by JelleZijlstra** [reply]
It doesn't necessarily give you an error; the type checker may just silently treat remaining code as unreachable. And any error you do get is going to be very unintuitive for a user.

**Reply by tkukushkin** [reply]
Ah, I get it, thank you. Waiting for python/typing#1043.
